In [ ]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [ ]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(columns=['EIN', 'NAME'])

In [ ]:
# Determine the number of unique values in each column.
unique_counts = application_df.nunique()
print(unique_counts)

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64


In [ ]:
# Look at APPLICATION_TYPE value counts for binning
application_counts = application_df['APPLICATION_TYPE'].value_counts()
print(application_counts)


APPLICATION_TYPE
T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: count, dtype: int64


In [ ]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
# Set a cutoff value for the number of occurrences for an application type to be considered "rare"

cutoff = 500

# Determine which values to replace
application_counts = application_df['APPLICATION_TYPE'].value_counts()
application_types_to_replace = list(application_counts[application_counts < cutoff].index)

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app, "Other")

# Check to make sure binning was successful
print(application_df['APPLICATION_TYPE'].value_counts())


APPLICATION_TYPE
T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: count, dtype: int64


In [ ]:
# Look at CLASSIFICATION value counts for binning
classification_counts = application_df['CLASSIFICATION'].value_counts()
print(classification_counts)


CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: count, Length: 71, dtype: int64


In [ ]:
# Look at CLASSIFICATION value counts > 1 for binning
classification_counts_gt1 = classification_counts[classification_counts > 1]
print(classification_counts_gt1)


CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: count, dtype: int64


In [ ]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
cutoff = 200

# Determine which values to replace
classification_counts = application_df['CLASSIFICATION'].value_counts()
classifications_to_replace = list(classification_counts[classification_counts < cutoff].index)

# Check if classifications_to_replace is empty
if classifications_to_replace:
    # Replace in dataframe
    for cls in classifications_to_replace:
        application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls, "Other")

   # Check to make sure binning was successful
print(application_df['APPLICATION_TYPE'].value_counts())


APPLICATION_TYPE
T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: count, dtype: int64


In [ ]:
# Convert categorical data to numeric with pd.get_dummies
application_df = pd.get_dummies(application_df)

# Display the first few rows of the modified dataframe
print(application_df.head())


   STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0       1     5000              1                   False   
1       1   108590              1                   False   
2       1     5000              0                   False   
3       1     6692              1                   False   
4       1   142590              1                   False   

   APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  \
0                  True                 False                False   
1                 False                 False                 True   
2                 False                 False                False   
3                 False                 False                 True   
4                 False                 False                 True   

   APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  ...  \
0                False                False                False  ...   
1                False                False                False  

In [ ]:


# Split our preprocessed data into our features and target arrays
y = application_df["IS_SUCCESSFUL"]
X = application_df.drop(columns=["IS_SUCCESSFUL"])

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

# Display the shapes of the training and testing sets
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)



(25724, 45) (8575, 45) (25724,) (8575,)


In [ ]:


# Create a StandardScaler instance
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)


## Compile, Train and Evaluate the Model

In [ ]:
# Define the model - deep neural net
number_input_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 100
hidden_nodes_layer2 = 50
hidden_nodes_layer3 = 20
hidden_nodes_layer4 = 10

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu"))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer3, activation="relu"))

# Fourth hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer4, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()


Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_5 (Dense)             (None, 100)               4600      
                                                                 
 dense_6 (Dense)             (None, 50)                5050      
                                                                 
 dense_7 (Dense)             (None, 20)                1020      
                                                                 
 dense_8 (Dense)             (None, 10)                210       
                                                                 
 dense_9 (Dense)             (None, 1)                 11        
                                                                 
Total params: 10891 (42.54 KB)
Trainable params: 10891 (42.54 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [ ]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])


In [ ]:
import os
from tensorflow.keras.callbacks import ModelCheckpoint

os.makedirs("checkpoints/", exist_ok=True)
checkpoint_path = "checkpoints/weights.{epoch:02d}.hdf5"
cp_callback = ModelCheckpoint(
    filepath=checkpoint_path,
    verbose=1,
    save_weights_only=True,
    save_freq=5)


In [ ]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=150, callbacks=[cp_callback])


Streaming output truncated to the last 5000 lines.
Epoch 135: saving model to checkpoints/weights.135.hdf5
444/804 [===============>..............] - ETA: 1s - loss: 0.5298 - accuracy: 0.7415
Epoch 135: saving model to checkpoints/weights.135.hdf5

Epoch 135: saving model to checkpoints/weights.135.hdf5

Epoch 135: saving model to checkpoints/weights.135.hdf5
459/804 [================>.............] - ETA: 1s - loss: 0.5291 - accuracy: 0.7421
Epoch 135: saving model to checkpoints/weights.135.hdf5

Epoch 135: saving model to checkpoints/weights.135.hdf5
470/804 [================>.............] - ETA: 1s - loss: 0.5294 - accuracy: 0.7420
Epoch 135: saving model to checkpoints/weights.135.hdf5

Epoch 135: saving model to checkpoints/weights.135.hdf5
483/804 [=================>............] - ETA: 1s - loss: 0.5302 - accuracy: 0.7408
Epoch 135: saving model to checkpoints/weights.135.hdf5

Epoch 135: saving model to checkpoints/weights.135.hdf5

Epoch 135: saving model to checkpoints/weig

In [44]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled, y_test, verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")


268/268 - 0s - loss: 0.5650 - accuracy: 0.7241 - 373ms/epoch - 1ms/step
Loss: 0.564978837966919, Accuracy: 0.7240816354751587


In [45]:
# Save the model to an HDF5 file
nn.save("AlphabetSoupCharity_Optimization.h5")

# Download the HDF5 file
from google.colab import files
files.download("AlphabetSoupCharity_Optimization.h5")


/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

# Alphabet Soup Deep Learning Model Performance Report

## Overview of the Analysis

The purpose of this analysis was to develop a deep learning model using TensorFlow and Keras to predict the success of funding applicants for Alphabet Soup based on various input features.

## Results

### Data Preprocessing

**Target Variable**: The target variable for the model is IS_SUCCESSFUL, which indicates whether the funding application was successful.

**Features**: The features for the model include APPLICATION_TYPE, AFFILIATION, CLASSIFICATION, USE_CASE, ORGANIZATION, STATUS, INCOME_AMT, and SPECIAL_CONSIDERATIONS.

**Removed Variables**: Variables EIN and NAME were removed from the input data as they were neither targets nor features.

### Compiling, Training, and Evaluating the Model

**Neurons, Layers, and Activation Functions**:
For the first model (AlphabetSoupCharity.ipynb), two hidden layers with 80 and 30 neurons were selected, respectively, using the ReLU activation function.
For the second model (AlphabetSoupCharity_Optimization.ipynb), the complexity was increased with four hidden layers having 100, 50, 20, and 10 neurons, respectively, using the ReLU activation function. The output layer for both models used the sigmoid activation function.

**Model Performance**:
The first model achieved an accuracy of 72.7%.
The second model achieved an accuracy of 72.4%.

**Target Model Performance**: While the target model performance of over 75% accuracy was not achieved for either model, with the first model achieving an accuracy of 72.7% and the second model achieving 72.4%, they provide a good starting point for further optimization.

**Steps Taken** :

Adjusting Binning:
Increased the cutoff value for binning rare occurrences in columns from 500 to 1000 in the second model, which did not significantly improve model accuracy.
Dropping Columns:
Explored dropping more or fewer columns from the input data, but the model's accuracy remained lower or similar.
Other Attempts:
Increased the number of hidden layers and neurons in the second model.
Used different activation functions for the hidden layers.
Increased the number of epochs to 120 and 170 in the second model.

**Steps to Increase Performance**:
Increased complexity with additional hidden layers and neurons in the second model.
Adjusted binning for rare occurrences in columns.
Increased the number of epochs to 200 in the second model.

## Summary

The deep learning models developed for Alphabet Soup aimed to predict the success of funding applicants based on various input features. The first model (AlphabetSoupCharity.ipynb) used two hidden layers with 80 and 30 neurons, achieving an accuracy of 72.7%. The second model (AlphabetSoupCharity_Optimization.ipynb) increased complexity with four hidden layers having 100, 50, 20, and 10 neurons, but achieved a slightly lower accuracy of 72.4%.

Despite not achieving the target model performance of over 75% accuracy, both models provided valuable insights. Adjustments were made in the second model, including increasing the cutoff value for binning rare occurrences, exploring different neural network architectures, and increasing the number of epochs to 200. These attempts aimed to improve model accuracy but did not yield significant improvements.

Moving forward, further optimization could involve fine-tuning hyperparameters, exploring more complex neural network architectures, and applying regularization techniques. Overall, while the models showed moderate success, there is room for improvement to enhance their predictive performance for funding application outcomes.





